community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/TaskCloser.scala

```diff
@@ -28,25 +28,29 @@ class TaskCloser {
    * @param task This task will be called, with true if the query went OK, and a false if an error occurred
    */
   def addTask(task: Boolean => Unit) {
+    if(closed)
+      throw new IllegalStateException("Already closed")
     _tasks += task
   }
 
   def close(success: Boolean) {
     if (!closed) {
       closed = true
-      val errors = _tasks.flatMap {
+      var foundException: Option[Throwable] = None
+      _tasks foreach {
         f =>
           try {
             f(success)
             None
           } catch {
-            case e: Throwable => Some(e)
+            case e: Throwable => foundException = Some(e)
           }
+
       }
 
-      errors.map(e => throw e)
+      foundException.forall(throwable => throw throwable)
```

Comment:
This changes semantics. Before, the first error in `errors` was thrown. In this, `foundException` gets overridden and actually the last encountered error is thrown. Is that a problem?